Convert flow types to TS

saihaj opened the pull request:
### Summary
Convert all the `.js` to `.ts` and fix all syntax related issues that exist because of Flow and TypeScript.

* [x] Fix `%checks` and remove `eslint-disable` comments (https://github.com/saihaj/graphql-js/commit/5d966bd052b692844e729cc83d84ee7ddea06205)
* [x] Change `_` https://github.com/saihaj/graphql-js/commit/fbbb7a557db0f5cd921a0d8b434bdad2c512c40c#r43625283
* [x] use `object` since it is a `WeakMap` (https://github.com/saihaj/graphql-js/commit/ae7c3f406e948ea7411acf02a2f88dc058428213#r43625385)
* [x] Restrict to `integrationTests/*/**` (https://github.com/saihaj/graphql-js/commit/28023e6bf172c63ed5ce00a102fb3c42684cb501)
* [x] Convert  `?` to `Maybe` (few were missed in 41b6975)
* [x] Convert `mixed` to `unknown` (few were missed in cd55ba9)
* [x] Convert to `$Keys` to `keyof`
* [x] Convert `$ReadOnly` to `Readonly`
* [x] Fixup caa8334 into c244b37
* [x] Complete comparison https://github.com/saihaj/graphql-js/pull/1 compare changes made manually with changes made by a tool https://github.com/Khan/flow-to-ts 
* [X] Replace `$ObjMap` to `ObjMap` utility function (need help on this. See discussion here https://github.com/graphql/graphql-js/pull/2828#discussion_r520238458)
* [X] Update `.babelrc-deno` to use TypeScript

Action items Based on https://github.com/graphql/graphql-js/pull/2828#issuecomment-761102728
For @IvanGoncharov 
* [x] Research if we can simplify `visit` typings in `16.x.x`
* [x] Remove polyfills
* [x] Fix test coverage on `16.x.x`
* [ ] https://github.com/graphql/graphql-js/pull/2828#discussion_r554121151
* [ ] Implement `GraphQLAgregatedError`
* [ ] Help @saihaj to generate declarations

For @saihaj 
* [x] Compare and remove .d.ts
* [x] https://github.com/graphql/graphql-js/pull/2828#discussion_r556184614
* [x] Remove Flow infrastructure, even integration test (remove in one commit and split if requested)
* [x] Convert type to interfaces
* [x] Enable as much as possible ESLint rules

~Fix no redeclare `src/polyfills/objectValues.ts`~ (these will be removed so no need)

joostheijkoop commented:
Hey! I was asked to check in here. You are doing some serious transformations. Is there any way I can help out? ;)

0558436961 reviewed (commented):
يتم اختراق حساب

IvanGoncharov commented on src/jsutils/isCollection.ts:
```diff
@@ -19,10 +19,7 @@
-
-// eslint-disable-next-line no-redeclare
-export default function isCollection(obj) {
+export default function isCollection(obj: unknown): obj is object| Symbol {
```
please squash it to `%checks` commit.

IvanGoncharov replied:
same for other similar changes.

saihaj commented on src/language/visitor.ts:
```diff
@@ -17,34 +18,34 @@ export type Visitor<KindToNode, Nodes = $Values<KindToNode>> =
     >;
-type EnterLeave<T> = {| +enter?: T, +leave?: T |};
+type EnterLeave<T> = { readonly enter?: T, readonly leave?: T };
 type ShapeMap<O, F> = $Shape<$ObjMap<O, F>>;
```
@IvanGoncharov I am confused on how to convert this`$ObjMap`?

joostheijkoop replied:
Do you mean this? https://github.com/Microsoft/TypeScript/issues/20352

saihaj replied:
Yes but I think there is a utility function for that in here https://github.com/graphql/graphql-js/blob/16.x.x/src/jsutils/ObjMap.js . I did try using that but it didn’t make sense  so I  am little confused

joostheijkoop replied:
It's seems to have one type parameter less, it might work better with de declaration from the linked issue. But need to dig deeper to investigate. Will try tomorrow

joostheijkoop replied:
how do you validate the syntax? I tried runnning `npx typescript` in the checkout of your branch, but to gave a lot (confusing) output

saihaj replied:
yes this is still work in progress and there are many issues that need to be resolved

IvanGoncharov replied:
@saihaj It's two different things:
* `$ObjMap` is Flow utility function https://flow.org/en/docs/types/utilities/#toc-objmap
* `ObjMap` is our utility type: https://github.com/graphql/graphql-js/blob/16.x.x/src/jsutils/ObjMap.js

IvanGoncharov replied:
So  please replace `$ObjMap` with TS equvivalent

saihaj replied:
> Do you mean this? [microsoft/TypeScript#20352](https://github.com/microsoft/TypeScript/issues/20352)

@IvanGoncharov so as @joostheijkoop mentioned that is the `$ObjMap` is TS equivalent. My questions are:
* Does this utility go in `src/jsutils`
  * Does this need any tests?
* What do I even name this utility since `ObjMap` is already a utility in codebase?
* Or should this change be handled in a new PR once we move `16.x.x` to `main`? all this conversation can be moved to an issue which can be handled by me or someone else in new PR?

IvanGoncharov commented:
@saihaj Can you please squash these two comments: 
https://github.com/graphql/graphql-js/pull/2828/commits/f47c55f3a878862f4215e35f081abf76ceb271f9
https://github.com/graphql/graphql-js/pull/2828/commits/c5ace4f2a134d1a80b5aaa125ee5979c20472f5c
Just use `git rebase -i` and move later commit earlier and squash them.
Please ping me if you need help with it.

saihaj commented:
@IvanGoncharov rebased this with `16.x.x.` and squashed those commits

IvanGoncharov commented on src/language/predicates.ts:
```diff
@@ -1,30 +1,41 @@
-import type { ASTNode } from './ast';
+import {
```
should be `import type`

IvanGoncharov commented on src/utilities/getIntrospectionQuery.ts:
```diff
@@ -254,25 +254,25 @@ export type IntrospectionNonNullTypeRef<
 export type IntrospectionTypeRef =
   | IntrospectionNamedTypeRef
-  | IntrospectionListTypeRef
+  | IntrospectionListTypeRef<any>
```
TS should support circular refs after 3.7, right?

saihaj replied:
Filed an issue https://github.com/microsoft/TypeScript/issues/42308 and also opened a question on [StackOverflow](https://stackoverflow.com/questions/65693844/using-recursive-type-aliases-gives-circular-reference-error) for help.

IvanGoncharov commented on src/type/definition.ts:
```diff
@@ -570,13 +570,12 @@ export class GraphQLScalarType {
     extensions: Maybe<ReadOnlyObjMap<unknown>>,
     extensionASTNodes: ReadonlyArray<ScalarTypeExtensionNode>,
-  } {
+  } & GraphQLScalarTypeConfig<unknown, unknown> {
```
change to inheritance after 998bea680d6e11e1c055a400a887a9539de08f75 is added to `16.x.x`

IvanGoncharov replied:
you should do that for `definition.ts` and `directives.ts`

IvanGoncharov commented on src/jsutils/PromiseOrValue.ts:
```diff
@@ -1 +1 @@
-export type PromiseOrValue<+T> = Promise<T> | T;
+export type PromiseOrValue<T> = Promise<T> | T;
```
Note to myself: I need to check this commit.

IvanGoncharov commented on tsconfig.json:
```diff
@@ -10,5 +10,6 @@
-  }
+  },
+  "exclude": ["integrationTests/*/**"]
 }
```
Superseed by 6e8ca04aff51a47f8378a234f4649f7aeb73514a please remove

IvanGoncharov commented on src/error/GraphQLError.ts:
```diff
@@ -83,7 +83,7 @@ export class GraphQLError extends Error {
     originalError?: Maybe<(Error & { readonly extensions?: unknown })>,
     extensions?: Maybe<{ [key: string]: unknown }>,
-  ): void {
+  ) {
```
Can you please extract this commit into PR against master?

saihaj replied:
@IvanGoncharov created a new PR https://github.com/graphql/graphql-js/pull/2878. I will remove this commit from here.

IvanGoncharov commented on src/language/visitor.ts:
```diff
@@ -45,7 +45,7 @@ export type VisitFn<TAnyNode, TVisitedNode extends TAnyNode = TAnyNode> = (
 export type VisitorKeyMap<KindToNode> = $ObjMap<
   KindToNode,
-  <T>(T) => ReadonlyArray<$Keys<T>>
+  <T>(T) => ReadonlyArray<keyof T>
```
Should go into next commit, right?

IvanGoncharov replied:
separate commit

IvanGoncharov commented on src/utilities/findBreakingChanges.ts:
```diff
@@ -65,12 +65,12 @@ export const DangerousChangeType = Object.freeze({
 
 export type BreakingChange = {
-  type: $Keys<typeof BreakingChangeType>;
+  type: keyof typeof BreakingChangeType;
```
same here

IvanGoncharov commented on src/language/blockString.ts:
```diff
@@ -93,8 +93,8 @@ export function getBlockStringIndentation(value: string): number {
-  indentation?: string = '',
-  preferMultipleLines?: boolean = false,
+  indentation: string = '',
+  preferMultipleLines: boolean = false,
```
Can you please extract it in PR against `master`?

saihaj replied:
@IvanGoncharov moved to https://github.com/graphql/graphql-js/pull/2879

IvanGoncharov commented on src/jsutils/__tests__/didYouMean-test.ts:
```diff
@@ -1,7 +1,7 @@
 import { describe, it } from 'mocha';
 
-import didYouMean from '../didYouMean';
+import { didYouMean } from '../didYouMean';
```
Can you please remove all default exports in separate PR against `16.x.x`?

saihaj replied:
Opened https://github.com/graphql/graphql-js/pull/2880 will remove this before I rebase

IvanGoncharov commented on src/execution/execute.ts:
```diff
@@ -60,12 +60,13 @@ import {
 import { getOperationRootType } from '../utilities/getOperationRootType';
 
+import { Maybe } from '../jsutils/Maybe';
+
```
Can you please review commit that introduces `Maybe` and fit it there?
Note: `jsutils` imports should be in the same group and should after `node_modules` imports.

IvanGoncharov commented:
Some action items for @IvanGoncharov:
* Research if we can simplify `visit` typings in `16.x.x`
* Remove polyfills
* Fix test coverage on `16.x.x`

Action for you @saihaj 
* Compare and remove `.d.ts`
* Remove Flow infrastructure, even integration test (remove in one commit and split if requested)
* Convert type to interfaces
* Enable as much as possible ESLint rules

IvanGoncharov commented on cspell.json:
```diff
@@ -46,6 +47,8 @@
     "Tatooine",
     "astromech",
+    "geocode",
+    "structs",
```
cspell have some issue with some dictionary marked as applicable on JS but not on TS.
For example:
https://github.com/streetsidesoftware/cspell-dicts/blob/6306b1d11d32e4f66721053d5fccf119ca7ce26e/dictionaries/fullstack/fullstack.txt#L46
https://github.com/streetsidesoftware/cspell-dicts/blob/6306b1d11d32e4f66721053d5fccf119ca7ce26e/dictionaries/fullstack/cspell-ext.json#L23

@saihaj Can you please contribute fixes to cspell?

IvanGoncharov commented on src/type/directives.ts:
```diff
@@ -189,16 +189,16 @@ export const GraphQLDeprecatedDirective = new GraphQLDirective({
 
 /**
- * Used to provide a URL for specifying the behaviour of custom scalar definitions.
+ * Used to provide a URL for specifying the behavior of custom scalar definitions.
```
same with `behaviour`

IvanGoncharov commented on src/polyfills/objectEntries.ts:
```diff
@@ -2,7 +2,6 @@ import type { ObjMap } from '../jsutils/ObjMap';
 declare function objectEntries<T>(obj: ObjMap<T>): Array<[string, T]>;
 
-/* eslint-disable no-redeclare */
 // $FlowFixMe[name-already-bound] workaround for: https://github.com/facebook/flow/issues/4441
```
Can you also remove Flow comments like this one?
Can be in a separate commit.

IvanGoncharov commented on src/error/GraphQLError.ts:
```diff
@@ -5,7 +5,7 @@ import type { Source } from '../language/source';
 import { getLocation } from '../language/location';
 import { printLocation, printSourceLocation } from '../language/printLocation';
-import { Maybe } from '../jsutils/Maybe';
+import type { Maybe } from '../jsutils/Maybe';
```
Please move it to commit that added `Maybe`

IvanGoncharov commented on src/error/formatError.ts:
```diff
@@ -34,14 +34,14 @@ export type GraphQLFormattedError = {
    * GraphQL document, it should contain a list of locations.
    */
-  readonly locations: ReadonlyArray<SourceLocation> | void;
+  readonly locations: ReadonlyArray<SourceLocation>;
```
Why did you remove `void`?
This props can have `undefined` as a value

IvanGoncharov commented on src/error/locatedError.ts:
```diff
@@ -1,5 +1,5 @@
 import { inspect } from '../jsutils/inspect';
-import { Maybe } from '../jsutils/Maybe';
+import type { Maybe } from '../jsutils/Maybe';
```
same

IvanGoncharov commented on src/error/locatedError.ts:
```diff
@@ -12,7 +12,7 @@ import { GraphQLError } from './GraphQLError';
 export function locatedError(
   rawOriginalError: unknown,
-  nodes: ASTNode | ReadonlyArray<ASTNode> | void | null,
+  nodes: ASTNode | ReadonlyArray<ASTNode> | null,
```
You can pass `undefined` why did you remove it here?

IvanGoncharov commented on src/execution/__tests__/abstract-test.ts:
```diff
@@ -42,6 +42,7 @@ async function executeQuery(args: {
 
 class Dog {
   name: string;
+
```
strange, why do you need empty line here?
what `eslint` rule is triggered by it?

saihaj replied:
I might have enabled incorrectly. Now should be updated

IvanGoncharov commented on src/execution/__tests__/abstract-test.ts:
```diff
@@ -52,6 +53,7 @@ class Dog {
 
 class Cat {
   name: string;
+
```
same here

IvanGoncharov commented on src/execution/__tests__/lists-test.ts:
```diff
@@ -37,6 +37,7 @@ describe('Execute: Accepts any iterable as list value', () => {
   });
 
   it('Accepts function arguments as a List value', () => {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
we had setting for it here: [`1ec501c` (#2828)](https://github.com/graphql/graphql-js/pull/2828/commits/1ec501c005fda0ca03614db8ad688a855ebbce0d#diff-9e1ecc14c733bb1ae2e523089f1262ac6ffccbcf950487ee0984403603550e57L630)
You need to remove 1ec501c from PR and just fix eslint errors here.

IvanGoncharov commented on src/execution/execute.ts:
```diff
@@ -262,7 +262,7 @@ export function buildExecutionContext(
   typeResolver?: Maybe<GraphQLTypeResolver<unknown, unknown>>,
 ): ReadonlyArray<GraphQLError> | ExecutionContext {
-  let operation: OperationDefinitionNode | void;
+  let operation: OperationDefinitionNode | undefined;
```
`void` => `undefined` should be a separate commit

saihaj commented:
I attempted to compare the `.d.ts` files will give one more go just to be sure and then remove them one

saihaj commented:
Compared and used PR #2899 from @kamilkisiela as a source to fix TS errors. I extracted a few things from that giant commit. That PR removed many `readonly` and there were some discrepancies in types compared to existing definitions which I didn't really understood so I just added `@ts-expect-error: FIXME` which can be fixed in future PRs

IvanGoncharov commented on src/error/__tests__/formatError-test.ts:
```diff
@@ -6,7 +6,6 @@ import { GraphQLError } from '../GraphQLError';
 
 describe('formatError: default error formatter', () => {
   it('uses default message', () => {
-    // $FlowExpectedError[incompatible-call]
```
@saihaj Please replace instances of `$FlowExpectedError` with `@ts-expect-error`.

IvanGoncharov commented on src/subscription/__tests__/subscribe-test.ts:
```diff
@@ -68,7 +68,7 @@ const EmailEventType = new GraphQLObjectType({
 const emailSchema = emailSchemaWithResolvers();
 
-function emailSchemaWithResolvers<T extends unknown>(
+function emailSchemaWithResolvers(
```
can you please locate the commit where you removed it and fix it there?

IvanGoncharov commented on src/jsutils/__tests__/inspect-test.ts:
```diff
@@ -169,7 +169,7 @@ describe('inspect', () => {
     expect(inspect([[new Foo()]])).to.equal('[[[Bar]]]');
 
-    const objectWithoutClassName = new (function () {
+    const objectWithoutClassName = new (function (this) {
```
it's incorrect fix, please return eslint disable comment you removed but change them to disable new rule

IvanGoncharov commented on src/__testUtils__/inspectStr.ts:
```diff
@@ -1,4 +1,4 @@
-import { Maybe } from '../jsutils/Maybe';
+import type { Maybe } from '../jsutils/Maybe';
```
please move this changes into PR where you added `Maybe`

IvanGoncharov commented:
@saihaj That was a long journey thanks a lot for your help 👍

pspeter3 commented:
Congrats! When will this be released?
